sdks/go/pkg/beam/core/funcx/output.go

```diff
@@ -84,6 +84,9 @@ func unfoldEmit(t reflect.Type) ([]reflect.Type, bool, error) {
 		if ok, err := isInParam(t.In(i)); !ok {
 			return nil, false, errors.Wrap(err, errIllegalParametersInEmit)
 		}
+		if t.In(i).Kind() == reflect.Interface || (t.In(i).Kind() == reflect.Pointer && t.In(i).Elem().Kind() == reflect.Interface) {
```

Comment:
Checking the Kind() against reflect.Interface is a good first approximation, but this prevents *all* interfaces from appearing in types in beam. 

The problem the issue needs to address isn't "no interfaces are allowed ever" but "the vanilla empty interface{}" isn't allowed.

But care needs to be taken: Not all "empty interfaces" are created equally. Since BeamGo was developed before Go Generics, we needed some way of simulating that kind of behavior. We called these "Universal Types".

Universal types are lightly described in the design RFC here:
https://s.apache.org/beam-go-sdk-design-rfc#heading=h.1vovlxln20bf

You can detect universal types with the function `typex.IsUniversal()`: https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/typex/class.go#L224

Look at how "IsUniversal" is defined, and how the different universal types are defined.

----
Further, we do allow arbitary interface types in these various function type signatures, but we require them to have a coder registered with Beam. That validation necessarily happens at a later stage, to avoid unnecessary coupling and complications at early parts of validation.

Since the empty interface can't have a coder defined for it (everything satisfies the empty interface), it's simpler to not allow it.
----

Then you can see that what we want to do for this bug is compare the type against the type of the empty interface.

```
var anyType = reflect.TypeOf((*any)(nil)).Elem()
```
(`any` is an alias for `interface{}`).

Let me know if you have questions about this.

Reply:
Thanks a lot! I'm clearer about this now.
Does it mean we should not allow empty interfaces like "interface{}", but we do allow 'Universal Types'?
But I still have a question is that, if a function has a param with "interface{}", like that
```
Fn := func(interface{}){}
t := reflect.TypeOf(Fn)
```
but typex.ClassOf(t.In(0)) returns "Concrete", I ought to think it will return "Universal".
The type system is a bit complex, I'm still learning it. :)

Reply:
Yes. Exactly. Deny "interface{}" but allow `beam.T` et al.

The "universal" types should only be referring to `beam.T` etc. Not arbitrary interfaces. Basically, it's important to distinguish between them (`beam.T` is not `beam.U`) so that for example, if a function is dropping a key or a value in a DoFn, we know which one is being used downstream./

Users can use beam.T and similar to escape the type system, but then they deserve any runtime type errors they get at pipeline execution time. The strictness of knowing what types are used where is to help users avoid expensive execution time issues.